### DISCUSSION TITLE
middleware.ts not running in `next dev` w/ nextjs 13 (but works with `next start`)

### DISCUSSION THREAD

**Discussion by MarkLyck** [vercel/next.js#46274] [Help] [answered]
I started a new project with NextJS 13 and appDir.

But I cannot get the middleware.ts file to work when in development mode. For some reason it runs fine if I do: `bun run build && bun run start` but if I just run my local development with `bun run dev` it doesn't work 🤷‍♂️

middleware.ts
```ts
import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'

export function middleware(request: NextRequest) {
  return NextResponse.redirect(new URL('/test', request.url))
}

export const config = {
  matcher: '/',
}
```

The file is placed in my root directory along side `next.config.js` and `app`

I tried adding `middleware.ts` to my tsconfig.json `include` array but that didn't make it work either.

This is on `next@13.1.6` I've also tried `next@canary` and it has the same problem.

Any idea how to solve this?

**Comment by icyJoseph** [comment]
I cannot reproduce this with `yarn`... could you try to discard that this is a bun issue?

Here's the file structures I tested with:

- without `src` dir:

```console
tree -I node_modules -L 3
.
├── README.md
├── app
│   ├── globals.css
│   ├── head.tsx
│   ├── layout.tsx
│   ├── page.module.css
│   ├── page.tsx
│   └── test
│       └── page.tsx
├── middleware.ts
├── next-env.d.ts
├── next.config.js
├── package.json
├── pages
│   └── api
│       └── hello.ts
├── public
│   ├── favicon.ico
│   ├── next.svg
│   ├── thirteen.svg
│   └── vercel.svg
├── tsconfig.json
└── yarn.lock

5 directories, 18 files
```

- with `src` dir:

```console
.
├── README.md
├── next-env.d.ts
├── next.config.js
├── package.json
├── public
│   ├── favicon.ico
│   ├── next.svg
│   ├── thirteen.svg
│   └── vercel.svg
├── src
│   ├── app
│   │   ├── globals.css
│   │   ├── head.tsx
│   │   ├── layout.tsx
│   │   ├── page.module.css
│   │   ├── page.tsx
│   │   └── test
│   ├── middleware.ts
│   └── pages
│       └── api
├── tsconfig.json
└── yarn.lock

6 directories, 16 files
```

**Reply by MarkLyck** [reply]
I tried with `yarn` as well and it's the same. Works when building a production bundle but does not work with `next dev`

My folder structure is the same as the first one above.

**Comment by MarkLyck** [comment, marked as answer]
Ahh it turns out it's a problem with `turbopack`

in my package.json my dev command is `next dev --turbo`

If I just run `bun run next dev` or remove the --turbo from package.json. The middleware will run properly in dev mode.

It's still in alpha as of writing this, so it's totally expected that not everything is working yet, but hopefully they add middleware support soon 🤞

There is already an issue open for this: https://github.com/vercel/next.js/issues/42921
